validate.go

```diff
@@ -1,16 +1,232 @@
+		if err := easyjson.Unmarshal(object.Request.Object, &replicationController); err != nil {
+			return nil, err
+		}
+		return replicationController.Metadata, nil
+	case CRONJOB_KIND:
+		cronjob := batchv1.CronJob{}
+		if err := easyjson.Unmarshal(object.Request.Object, &cronjob); err != nil {
+			return nil, err
+		}
+		return cronjob.Metadata, nil
+	case JOB_KIND:
+		job := batchv1.Job{}
+		if err := easyjson.Unmarshal(object.Request.Object, &job); err != nil {
+			return nil, err
+		}
+		return job.Metadata, nil
+	case POD_KIND:
+		pod := corev1.Pod{}
+		if err := easyjson.Unmarshal(object.Request.Object, &pod); err != nil {
+			return nil, err
+		}
+		return pod.Metadata, nil
+	default:
+		return nil, fmt.Errorf("object should be one of these kinds: Deployment, ReplicaSet, StatefulSet, DaemonSet, ReplicationController, Job, CronJob, Pod. Found %s", object.Request.Kind.Kind)
+	}
+}
+
+func updateLabels(labels *map[string]string, newLabels map[string]string) bool {
+	has_mutation := false
+	for label, newValue := range newLabels {
+		if oldValue, has_label := (*labels)[label]; !has_label || oldValue != newValue {
+			(*labels)[label] = newValue
+			has_mutation = has_mutation || true
+		}
+	}
+	return has_mutation
+}
+
+func updateResourceLabels(object kubewarden_protocol.ValidationRequest, newLabels map[string]string) ([]byte, error) {
+	switch object.Request.Kind.Kind {
+	case DEPLOYMENT_KIND:
+		deployment := appsv1.Deployment{}
+		if err := easyjson.Unmarshal(object.Request.Object, &deployment); err != nil {
+			return nil, err
+		}
+		if deployment.Metadata.Labels == nil {
+			deployment.Metadata.Labels = make(map[string]string)
+		}
+		if updateLabels(&deployment.Metadata.Labels, newLabels) {
+			return kubewarden.MutateRequest(deployment)
+		}
+		return kubewarden.AcceptRequest()
+	case REPLICASET_KIND:
+		replicaset := appsv1.ReplicaSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &replicaset); err != nil {
+			return nil, err
+		}
+		if replicaset.Metadata.Labels == nil {
+			replicaset.Metadata.Labels = make(map[string]string)
+		}
+		if updateLabels(&replicaset.Metadata.Labels, newLabels) {
+			return kubewarden.MutateRequest(replicaset)
+		}
+		return kubewarden.AcceptRequest()
+	case STATEFULSET_KIND:
+		statefulset := appsv1.StatefulSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &statefulset); err != nil {
+			return nil, err
+		}
+		if statefulset.Metadata.Labels == nil {
+			statefulset.Metadata.Labels = make(map[string]string)
+		}
+		if updateLabels(&statefulset.Metadata.Labels, newLabels) {
+			return kubewarden.MutateRequest(statefulset)
+		}
+		return kubewarden.AcceptRequest()
+	case DAEMONSET_KIND:
+		daemonset := appsv1.DaemonSet{}
+		if err := easyjson.Unmarshal(object.Request.Object, &daemonset); err != nil {
+			return nil, err
+		}
+		if daemonset.Metadata.Labels == nil {
+			daemonset.Metadata.Labels = make(map[string]string)
+		}
+		if updateLabels(&daemonset.Metadata.Labels, newLabels) {
```

Comment:
in addition to that, I think we should also look for the labels of the PodSpec and ensure the final Pod has the relevant labels too